### PR TITLE
Implement active organization.

### PR DESCRIPTION
**Pull request by MGLL**
Hello there, 
if you are still interested into "active organization" feature, here is a contribution.

## Summary
- Implemented "Active organization" based on user's attributes.
- Created mappers for active organization data.
- Created endpoint to be able to switch organization, it should return an updated access and refresh token.
- Created endpoint to be able to get the active organization info.
- Implemented tests in "testOrganizationSwitch()"
- Extended AbstractOrganizationTest functionalities for advanced testing.
- Created documentation in "active-organization.md"
- Created some util classes (ActiveOrganizationUtil, TokenManagerUtil)

## Details
_(more info into active-organization.md)_

### Endpoints

I have created 2 new endpoints:
- `PUT /:realm/users/switch-organization`
  - with organization id in request body

This endpoint assigns the organization's id to the user's attributes under the `org.ro.active` key.
Before assigning, it checks that the target organization exists, verify that the user is a member of it.

The endpoint returns a 200 "OK" response with an updated access_token and refresh_token similar to the authentication flow.

- `GET /:realm/users/active-organization`

This endpoint returns the active organization information (after standard verification: that the organization exist, and the user is a member of it)

### Mappers

I have added fine grained ActiveOrganization Mappers, so everyone can adjust following its taste.
- ActiveOrganizationMapper
- ActiveOrganizationInfoMapper
- ActiveOrganizationAttributeMapper
- ActiveOrganizationRoleMapper

## Next steps?
- Probably updating https://github.com/MGLL/phasetwo-docs to include those new endpoints, if approved by you.

## Planned evolutions
Next iteration (later when I have more time):
- Implement external configuration (through API) for active organization attribute key.
- Detect and adapt implementation based "declarative user profile" preview enabled feature (to create a "declarative" read only "User profile attributes" by default like email or username).
- Implement an alternative "active organization" mode to rely on session instead of attributes (like https://github.com/p2-inc/keycloak-orgs/issues/149).
  - (and) Implement external configuration (through API) to switch active organization mode "SESSION" or "ATTRIBUTES".

## End

If you have any feedbacks or request for modifications, let me know, I will adjust it.

Best regards

### CODE REVIEW

**Comment by MGLL** [comment]
@xgp You are welcome!

No problem, I will fix everything tomorrow after work or Friday at worst.

For 3. I wasn't aware it was possible, if I knew it, I would have did it directly 😄. Thanks for pointing it out and actually after a quick test, I think I could do something.

**However**, it seems there is a bug in Keycloak for `MULTIVALUED_STRING_TYPE` and `MULTIVALUED_LIST_TYPE`, where creating the mapper send the info in an Array, but Keycloak endpoint `protocol-mappers/models` doesn't expect it.

Which isn't great as you can't do a nice multivalued drop-down...

Config:
image

Request Body:
image

Error log:
image

If I use "STRING" type instead as a workaround and rely on tooltip (and documentation) to help users:
image

image

It then works
image

I could then split the received String and apply some logic. But it doesn't look nice for the end user. 

I will raise this issue in Keycloak tomorrow, but looking at the source code, I'm not sure there will be a fix, as config in `ProtocolMapperRepresentation` expect a `Map<String, String>`.
image


However, if I progress with a configurable mapper, please note that I will need to get access to / use `ProtocolMapperModel mappingModel` to get the inputed values (so I might need to override the `setClaim` in `ActiveOrganizationMapper` or change `getOrganizationClaim`). 

`Side note: if you have other picky comment, don't hesitate, I'm always looking to improve.`

File: src/main/java/io/phasetwo/service/util/Constants.java
[the remark this answers is not in the document]
Comment (MGLL) [reply src/main/java/io/phasetwo/service/util/Constants.java:1]: `Constants.java` was deleted and the constant moved into `Orgs.java`

File: active-organization.md
[the remark this answers is not in the document]
Comment (MGLL) [reply active-organization.md:1]: Documentation was moved (_and adjusted_) under `docs/` and a link was provided from the README.

File: src/main/java/io/phasetwo/service/protocol/oidc/mappers/ActiveOrganizationAttributeMapper.java
[the remark this answers is not in the document]
Comment (MGLL) [reply src/main/java/io/phasetwo/service/protocol/oidc/mappers/ActiveOrganizationAttributeMapper.java:1]: I created a single mapper which is configurable. 
Sadly, it seems that mapper doesn't support multi-valued input which is sent as Array.
So, I did a configurable mapper based on string input with a default configuration setup (see doc).

Opened a [discussion](https://github.com/keycloak/keycloak/discussions/26314#discussion-6091445) on that.

File: src/main/java/io/phasetwo/service/util/ActiveOrganizationUtil.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.jboss.logging.Logger;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+
+public class ActiveOrganizationUtil {
```
[the remark this answers is not in the document]
Comment (MGLL) [reply src/main/java/io/phasetwo/service/util/ActiveOrganizationUtil.java]: Fixed

File: src/main/java/io/phasetwo/service/util/TokenManagerUtil.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.representations.AccessTokenResponse;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.AuthenticationSessionManager;
+import org.keycloak.services.util.MtlsHoKTokenUtil;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.sessions.RootAuthenticationSessionModel;
+
+public class TokenManagerUtil {
```
[the remark this answers is not in the document]
Comment (MGLL) [reply src/main/java/io/phasetwo/service/util/TokenManagerUtil.java]: Fixed

**Comment by MGLL** [comment]
@xgp hello, this PR is ready for a new review.

**Comment by MGLL** [comment]
Hi @xgp 

I see it's still in pending, I think I have solved all requested change.
If I missed something, let me know. 

Best regards,

**Comment by xgp** [comment]
@MGLL No problems. The issue is just time. We'll get to it soon.

**Comment by MGLL** [comment]
> @MGLL No problems. The issue is just time. We'll get to it soon.

Thanks for info, no Problem, let me know if there is anything 👍

**Comment by xgp** [comment]
@MGLL I was able to review again. Thanks for making the changes.

One more thing that I missed in the first review is what happens when the Organization is deleted or the User is removed from the Organization. We should remove the attribute in those cases. I see the check that makes sure the User is a member of an Organization before allowing the claim mapping, but I think it would be good style to also explicitly remove the attribute.

Other than that, I think it's almost ready to go.

**Comment by MGLL** [comment]
> One more thing that I missed in the first review is what happens when the Organization is deleted or the User is removed from the Organization. 

Ah damn, good catch. 
I will fix this (I will add a check in the logic where a user is removed from an organization to remove it from attribute if it's the active one).

**Comment by phamann** [comment]
Really happy to see this addition @MGLL @xgp!

I'm also happy you went with the user attributes route instead of the session notes as originally proposed in one of the linked issues. As session notes do not persist across tab's in Keycloak. 

Coincidentally, we've implemented something similar using `keycloak-orgs` in our private build of Keycloak; however, it has a bit of additional business logic unique to our needs and the legacy domain model we were transitioning from.

One thing I thought I'd suggest/note: in our implementation, instead of using a custom endpoint like `PUT /:realm/users/switch-organization` for user org switching, we opted for a more spec-compliant way using a customer authenticator which acted upon the presence of the OIDC query parameter `promp=select_account`. 

From the [spec](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest):

> `select_account`
>
>    The Authorization Server SHOULD prompt the End-User to select a user account. This enables an End-User who has multiple accounts at the Authorization Server to select amongst the multiple accounts that they might have current sessions for. If it cannot obtain an account selection choice made by the End-User, it MUST return an error, typically account_selection_required. 


This way, we could use Keycloak's existing auth and token endpoints and off-the-shelf OIDC clients without any modification for account switching. We also added an `account_hint=<org_id>` param, allowing users to switch directly to an account without being challenged to choose from a list, similar to Keycloaks `kc_idp_hint` behaviour.

That being said, such an authenticator could co-exist with this implementation and endpoint. But maybe something for you to consider in the future.

**Comment by MGLL** [comment]
@phamann Thanks for the feedback.

From what I understood, it would require creating an Authenticator like 
```java
ActiveOrganizationAuthenticator implements Authenticator {...}
```

Which is displayed in the authentication flow when the `prompt=select_account` param is given to the keycloak /token endpoints. And here, you select an organization for the "session" similar to that:
image

But if the user has no organization, it should fall into an error, according to 
> If it cannot obtain an account selection choice made by the End-User, it MUST return an error, typically account_selection_required.

**That something I could work on in a next PR**, as it requires more time for me to do that*, and this implementation already kind of solve a pain point (_note that I took example on another solution which provide this kind of endpoints_). 
*Basically, I'm currently missing some knowledge to know how to do that which is: 
- How can I catch in the authentication request that the `prompt=select_account` was given. 

If you can already give me some hints, it could save me a bit of time. 

Then I guess I have to use something like
```java
ActiveOrganizationAuthenticatorCondition implements ConditionalAuthenticator {...}
```
To decide if I should display `ActiveOrganizationAuthenticator` or not. 

I also like a lot the `account_hint=<org_id> param` which avoid asking the user to select an organization at each log-in (which add effort on the user side). **But**, would it mean that we have to "override" / "customize" the Keycloak /token endpoints? (so, we can add this params).

Also, this point
> If it cannot obtain an account selection choice made by the End-User, it MUST return an error, typically account_selection_required.

Could be a problem for our use case (and others), as for example, organization can be optional in a SaaS Platform. For example, some users won't be part of an organization and some will be part of 1 and some will be part of multiple.
So, if I do this change in another PR, I will try to add an external configuration to decide if the organization selection should be mandatory (and then "` return an error, typically account_selection_required`") or optional (allow the user to proceed to the platform).

**Comment by MGLL** [comment]
@xgp I have added some changes to remove the active org attribute when the membership is removed (and if it was the active organization).
However, regarding the 
> Organization is deleted

Should I add the changes too? 
Because from what I see, it would require to
- get all members of the organization
- iterate over them
- verify if the active organization attribute matches the deleted organization
- if yes, remove the attribute

Which I fear might be a bit extensive depending of the number of members.
If this is fine for you, I will add it. 
Else, `ActiveOrganization.isValid()` should deal with it as it verifies the membership and if the organization exists. 

Let me know

**Comment by xgp** [comment]
> Which I fear might be a bit extensive depending of the number of members.

Agreed. Is there a good place to put in a lazy evaluation of that? I.e. if we come across an "active org" user attribute for a nonexistent org, can we remove it then?

Otherwise, I'm ready to merge this. Thanks again for all of the great work on this @MGLL (and patience in waiting for me to review).

**Comment by MGLL** [comment]
Hmm we could check "_during login_" ? for example, through the `ActiveOrganization` utility when the `ActiveOrganizationMapper` calls the `isValid` method. 

Because, this method check here if the user is a member of the organization and return _false_ if it's not the case. I could add an additional check there to verify if the organization still exists (and remove the user's attribute if not) before returning the false state.  
The downside is that it makes an additional database access, but actually it would makes sense. Because, if you land in this condition, the organization might not exists anymore but you still have the active org attribute for this one. 

What do you think ?

**Comment by xgp** [comment]
> What do you think ?

That sounds like a good place to do the check. Thanks.

**Comment by MGLL** [comment]
@xgp Did the change.

Added
```java
      // verify that the organization still exists
      organization = organizationProvider.getOrganizationById(realm, activeOrganizationId);
      if (organization == null) {
        log.warnf("organization doesn't exists anymore.");
        user.removeAttribute(ACTIVE_ORGANIZATION);
      }
```

in ActiveOrganization utility (in `isValid`)

Updated test with
```java
    // delete organization
    deleteOrganization(org3Id);

    // re-authenticate
    kc = getKeycloak(REALM, "test-ui", "user", "password");

    // verify attribute is removed
    validateActiveOrganizationFromUserAttributes(getUserAccount(kc), false, null);
```

to cover the case

**Review by xgp — APPROVED** [review]

**Pull request merged**